### PR TITLE
Mpi particlefilter

### PR DESCRIPTION
**Pull request by babrodtk**

### CODE REVIEW

File: gpu_ocean/SWESimulators/IPythonMagic.py [hunk 1]
```diff
@@ -76,30 +88,69 @@ def exitfunc():
             self.logger.debug("==================================================================")
         atexit.register(exitfunc)
         
+        
+        
+        
+        
+        
+    logger_initialized = False
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/IPythonMagic.py]: This makes sure that if you use logging, but run the initialization magic twice, it only initializes once. Very handy

File: gpu_ocean/SWESimulators/IPythonMagic.py [hunk 2]
```diff
@@ -76,30 +88,69 @@ def exitfunc():
...
+
+            #Add log to screen
+            ch = logging.StreamHandler()
+            ch.setLevel(args.level)
+            logger.addHandler(ch)
+            logger.log(args.level, "Console logger using level %s", logging.getLevelName(args.level))
+            
+            #Get the outfilename (try to evaluate if Python expression...)
+            try:
+                outfile = eval(args.out, self.shell.user_global_ns, self.shell.user_ns)
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/IPythonMagic.py]: This enables using python statements as log filenames - very handy for e.g., MPI

File: gpu_ocean/SWESimulators/MPIOceanModelEnsemble.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
+# -*- coding: utf-8 -*-
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/MPIOceanModelEnsemble.py]: This class manages MPI communication between nodes essentially

File: gpu_ocean/SWESimulators/OceanModelEnsemble.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/OceanModelEnsemble.py]: Runs an ensemble on a single node

File: gpu_ocean/SWESimulators/gpu_kernels/ocean_noise.cu [hunk 5]
```diff
@@ -25,10 +25,10 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
   *  Generates two uniform random numbers based on the ANSIC Linear Congruential 
   *  Generator.
   */
-__device__ float2 ansic_lcg(unsigned long* seed_ptr) {
-    unsigned long seed = (*seed_ptr);
+__device__ float2 ansic_lcg(unsigned long long* seed_ptr) {
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/gpu_kernels/ocean_noise.cu]: Bugfix here: unsigned long is 32 bits, whilst unsigned long long is 64 bits

File: gpu_ocean/demos/sequential_importance_resampling.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
+# -*- coding: utf-8 -*-
```
Comment (babrodtk) [inline gpu_ocean/demos/sequential_importance_resampling.py]: Python script that should be able to run SIR on a cluster of nodes using MPI. Look at mainLoop for what is actually going on . Most of the other functions are simply setup of initial conditions etc. and plotting.

**Comment by setmar** [comment]
Latest: Completede successful run with 8 nodes on Abel. Setting number of particles to 10 or 100 results in job timeout. Race condition?

Will merge PR now, but we have to look into the possible race condition.

**Pull request merged**